### ISSUE TITLE
Debugger: std::maps where the key is a class are not displayed correctly

### ISSUE THREAD

**Issue by tilman22** [microsoft/vscode-cpptools#835]
Hi,

please see the following test code:
```
#include <map>

class CInt
{
  public:
    CInt(int ii = -1) { i = ii; }
    bool operator < (const CInt& other) const { return this->i < other.i; }
    int i;
};

int main(void)
{
  std::map<int, CInt> intmap2;
  intmap2[3] = 1;
  intmap2[1] = 2;
  intmap2[7] = 3;
  intmap2[5] = 4;

  std::map<CInt, CInt> intmap3;
  intmap3[3] = 1;
  intmap3[1] = 2;
  intmap3[7] = 3;
  intmap3[5] = 4;

  return 0;
}
```

The key of the second map (intmap3) is a class. Instances of this map can be pretty-printed by gdb correctly:
```
-exec p intmap2
$1 = std::map with 4 elements = {[1] = {i = 2}, [3] = {i = 1}, [5] = {i = 4}, [7] = {i = 3}}
-exec p intmap3
$2 = std::map with 4 elements = {[{i = 1}] = {i = 2}, [{i = 3}] = {i = 1}, [{i = 5}] = {i = 4}, [{i = 7}] = {i = 3}}

```
But this instance intmap3 is not displayed by vscode/cpp-extension correctly:
[https://www.dropbox.com/s/57g077p5hpt42cz/png.png?dl=0](url)

https://www.dropbox.com/s/57g077p5hpt42cz/png.png?dl=0

Tilman

**Comment by tilman22** [comment]
We are using a lot of std::maps<TKey, TValue> where TKey is not a primitive but a class. It is annyoing (remark: all in all, vscode is the very best editor I ever worked with!) that the debugger is not able to output the key-value.

But there is a workaround: In the printer.py file, change the display-hint from "map" to "array" (or comment out this function at all). At least, you will be able to see the keys (with even indices).
```
class StdMapPrinter:
    "Print a std::map or std::multimap"
   ...
    def display_hint (self):
        return 'array'

```

**Comment by pieandcakes** [comment]
@tilman22 this sounds like a gdb pretty-printer issue. The debugger extension is not doing additional work to visualize the data. Instead, we rely on gdb to return the data in a format that can be visualized.

If you enable `"logging": { "engineLogging": true }` you can see what we request and what gdb returns with and without your change. This might be a good item to add to the documentation in this repo for other users running into the same visualization issue.

**Comment by tilman22** [comment]
Hi pieandcakes,
for me it sounds like a parsing problem of the debugging extension.  Parsing only seems to work when the key of the map is a primitive, not a class.
This is the logging-output:
```
1: (54425) ->1043^done,name="var5",numchild="0",value="{...}",type="std::map<int, CInt, std::less<int>, std::allocator<std::pair<int const, CInt> > >",thread-id="1",displayhint="map",dynamic="1",has_more="1"
1: (54425) ->(gdb)
1: (54425) 1043: elapsed time 1
1: (54425) <-1044-var-create - * "intmap3"
1: (54426) ->1044^done,name="var6",numchild="0",value="{...}",type="std::map<CInt, CInt, std::less<CInt>, std::allocator<std::pair<CInt const, CInt> > >",thread-id="1",displayhint="map",dynamic="1",has_more="1"
1: (54426) ->(gdb)
1: (54426) 1044: elapsed time 1
1: (54431) <-1045-var-list-children --simple-values "var5" 0 1000

here: inspecting the map which has an integer as the key:

1: (54433) ->1045^done,numchild="8",displayhint="map",children=[child={name="var5.[0]",exp="[0]",numchild="0",value="1",type="const int",thread-id="1"},child={name="var5.[1]",exp="[1]",numchild="1",type="CInt",thread-id="1"},child={name="var5.[2]",exp="[2]",numchild="0",value="3",type="const int",thread-id="1"},child={name="var5.[3]",exp="[3]",numchild="1",type="CInt",thread-id="1"},child={name="var5.[4]",exp="[4]",numchild="0",value="5",type="const int",thread-id="1"},child={name="var5.[5]",exp="[5]",numchild="1",type="CInt",thread-id="1"},child={name="var5.[6]",exp="[6]",numchild="0",value="7",type="const int",thread-id="1"},child={name="var5.[7]",exp="[7]",numchild="1",type="CInt",thread-id="1"}],has_more="0"
1: (54433) ->(gdb)
1: (54433) 1045: elapsed time 2
1: (54433) <-1046-var-list-children --simple-values "var5.[1]" 0 1000
1: (54434) ->1046^done,numchild="1",children=[child={name="var5.[1].public",exp="public",numchild="1",value="",thread-id="1"}],has_more="0"
1: (54434) ->(gdb)
1: (54434) 1046: elapsed time 1
1: (54434) <-1047-var-list-children --simple-values "var5.[1].public" 0 1000
1: (54435) ->1047^done,numchild="1",children=[child={name="var5.[1].public.i",exp="i",numchild="0",value="2",type="int",thread-id="1"}],has_more="0"
1: (54435) ->(gdb)
1: (54435) 1047: elapsed time 1
1: (54435) <-1048-var-list-children --simple-values "var5.[3]" 0 1000
1: (54436) ->1048^done,numchild="1",children=[child={name="var5.[3].public",exp="public",numchild="1",value="",thread-id="1"}],has_more="0"
1: (54436) ->(gdb)
1: (54436) 1048: elapsed time 1
1: (54436) <-1049-var-list-children --simple-values "var5.[3].public" 0 1000
1: (54437) ->1049^done,numchild="1",children=[child={name="var5.[3].public.i",exp="i",numchild="0",value="1",type="int",thread-id="1"}],has_more="0"
1: (54437) ->(gdb)
1: (54437) 1049: elapsed time 1
1: (57239) <-1050-var-list-children --simple-values "var6" 0 1000

here: inspecting the map which has a class (CInt) as the key:

1: (57247) ->1050^done,numchild="8",displayhint="map",children=[child={name="var6.[0]",exp="[0]",numchild="1",type="const CInt",thread-id="1"},child={name="var6.[1]",exp="[1]",numchild="1",type="CInt",thread-id="1"},child={name="var6.[2]",exp="[2]",numchild="1",type="const CInt",thread-id="1"},child={name="var6.[3]",exp="[3]",numchild="1",type="CInt",thread-id="1"},child={name="var6.[4]",exp="[4]",numchild="1",type="const CInt",thread-id="1"},child={name="var6.[5]",exp="[5]",numchild="1",type="CInt",thread-id="1"},child={name="var6.[6]",exp="[6]",numchild="1",type="const CInt",thread-id="1"},child={name="var6.[7]",exp="[7]",numchild="1",type="CInt",thread-id="1"}],has_more="0"
1: (57247) ->(gdb)
1: (57247) 1050: elapsed time 8

```

Best regards,
Tilman

**Comment by pieandcakes** [comment]
@tilman22 This should be fixed in `0.12.1` which was released yesterday. Can you see if this is still an issue?

**Comment by pieandcakes** [comment]
Wrong issue
